File: src/mk3/JavaPackage/src/main/java/tamaf/AgentID.java
package tamaf;

public class AgentID {
    private final String name;
    private Address address;

    public AgentID(String name) {
        this.name = name;
    }

    public AgentID(String name, Address address) {
        this.name = name;
        this.address = address;
    }

    public String getName() { return name; }
    public Address getAddress() { return address; }

    public void setPort(int port) {
        if (this.address != null) {
            this.address.setPort(port);
        }
    }

    public String getFullID() {
        if (address == null) return name + "@None:None";
        return name + "@" + address.getIp() + ":" + (address.getPort() != null ? address.getPort() : "None");
    }

    public static AgentID fromString(String aidString) {
        String[] parts = aidString.split("@", 2);
        String name = parts[0];
        String networkPart = parts[1];
        
        if (name.equals("None")) name = null;
        
        int lastColon = networkPart.lastIndexOf(':');
        String ip = networkPart.substring(0, lastColon);
        String portStr = networkPart.substring(lastColon + 1);
        
        if (ip.equals("None")) {
            return new AgentID(name, null);
        }
        
        Integer port = portStr.equals("None") ? null : Integer.valueOf(portStr);
        return new AgentID(name, new Address(ip, port));
    }

    public static boolean matches(AgentID template, AgentID target) {
        if (target == null) return false;
        
        // If template has a name, it must match.
        // In this framework, name@address is the full ID, but name should be unique enough for most templates.
        if (template.getName() != null) {
            if (template.getName().equals(target.getName())) {
                // Name matches. If address is also provided in template, we can check it,
                // but for cross-platform/multi-IP stability, we should be lenient if only name was intended.
                // If the template has NO address, we consider it a name-only match.
                if (template.getAddress() == null) {
                    return true;
                }
                
                // If template HAS address, we check it.
                if (target.getAddress() == null) return false;
                
                if (template.getAddress().getIp() != null && !template.getAddress().getIp().equals(target.getAddress().getIp())) {
                    return false;
                }
                
                if (template.getAddress().getPort() != null && !template.getAddress().getPort().equals(target.getAddress().getPort())) {
                    return false;
                }
                
                return true;
            } else {
                return false;
            }
        }
        
        // If name is null in template, we must match by address
        if (template.getAddress() != null) {
            if (target.getAddress() == null) return false;
            
            if (template.getAddress().getIp() != null && !template.getAddress().getIp().equals(target.getAddress().getIp())) {
                return false;
            }
            
            if (template.getAddress().getPort() != null && !template.getAddress().getPort().equals(target.getAddress().getPort())) {
                return false;
            }
            return true;
        }

        // Empty template matches everything
        return true;
    }
}

File: src/mk1/PythonPackage/src/tamaf/agentid.py
from .address import Address

class AgentID:
    
    def __init__(self, name: str, address: Address = None):
        self.name = name
        self.address = address

    def SetPort(self, port: int):
        if self.address:
            self.address.SetPort(port)

    def GetFullID(self) -> str:
        if self.address:
            return f"{self.name}@{self.address.ip}:{self.address.port}"
        return f"{self.name}@None:None"
    
    def GetIDNoPort(self) -> str:
        if self.address:
            return f"{self.name}@{self.address.ip}"
        return f"{self.name}@None"

    @classmethod
    def FromString(cls, aidString: str):
        name, networkPart = aidString.split('@', 1)
        ip, portString = networkPart.rsplit(':', 1)

        if name == "None":
            name = None

        if ip == "None":
            return cls(name=name, address=None)

        if portString != "None":
            port = int(portString)
            return cls(name=name, address=Address(ip=ip, port=port))
        else:
            return cls(name=name, address=Address(ip=ip))
            
    @staticmethod
    def Matches(template: 'AgentID', target: 'AgentID'):
        if target is None:
            return False
            
        # If template has a name, it must match.
        if template.name is not None:
            if template.name == target.name:
                # Name matches. If address is also provided in template, we check it.
                if template.address is None:
                    return True
                
                # If template HAS address, we check it.
                if target.address is None:
                    return False
                
                if template.address.ip is not None and template.address.ip != target.address.ip:
                    return False
                
                if template.address.port is not None and template.address.port != target.address.port:
                    return False
                
                return True
            else:
                return False
        
        # If name is None in template, we must match by address
        if template.address is not None:
            if target.address is None:
                return False
            
            if template.address.ip is not None and template.address.ip != target.address.ip:
                return False
            
            if template.address.port is not None and template.address.port != target.address.port:
                return False
            
            return True

        # Empty template matches everything
        return True
